src/equations/presetEquationRunnerWASM.js

```diff
@@ -0,0 +1,162 @@
+import Utils from '../utils';
+
+export default class PresetEquationRunnerWASM {
+  constructor (preset, globalVars, opts) {
+    this.preset = preset;
+
+    this.texsizeX = opts.texsizeX;
+    this.texsizeY = opts.texsizeY;
+    this.mesh_width = opts.mesh_width;
+    this.mesh_height = opts.mesh_height;
+    this.aspectx = opts.aspectx;
+    this.aspecty = opts.aspecty;
+    this.invAspectx = 1.0 / this.aspectx;
+    this.invAspecty = 1.0 / this.aspecty;
+
+    this.qs = Utils.range(1, 33).map((x) => `q${x}`);
+    this.ts = Utils.range(1, 9).map((x) => `t${x}`);
+    this.regs = Utils.range(100).map((x) => {
+      if (x < 10) {
+        return `reg0${x}`;
+      }
+      return `reg${x}`;
+    });
+
+    this.initializeEquations(globalVars);
+  }
+
+  initializeEquations (globalVars) {
+    this.runVertEQs = (this.preset.pixel_eqs !== '');
+
+    this.mdVSQInit = null;
+    this.mdVSRegs = null;
+    this.mdVSFrame = null;
+    this.mdVSUserKeys = null;
+    this.mdVSFrameMap = null;
+
+    this.mdVSShapes = null;
+    this.mdVSUserKeysShapes = null;
+    this.mdVSFrameMapShapes = null;
+
+    this.mdVSWaves = null;
+    this.mdVSUserKeysWaves = null;
+    this.mdVSFrameMapWaves = null;
+
+    this.mdVSQAfterFrame = null;
+
+    // this.gmegabuf = new Array(1048576).fill(0);
+
+    const mdVSBase = {
+      frame: globalVars.frame,
+      time: globalVars.time,
+      fps: globalVars.fps,
+      bass: globalVars.bass,
+      bass_att: globalVars.bass_att,
+      mid: globalVars.mid,
+      mid_att: globalVars.mid_att,
+      treb: globalVars.treb,
+      treb_att: globalVars.treb_att,
+      meshx: this.mesh_width,
+      meshy: this.mesh_height,
+      aspectx: this.invAspectx,
+      aspecty: this.invAspecty,
+      pixelsx: this.texsizeX,
+      pixelsy: this.texsizeY,
+      // gmegabuf: this.gmegabuf,
+    };
+
+    this.mdVS = Object.assign({}, this.preset.baseVals, mdVSBase);
+
+    Utils.setWasm(this.preset.globals, mdVSBase, Object.keys(mdVSBase));
+
+    // this.mdVS.megabuf = new Array(1048576).fill(0);
+    this.rand_start = new Float32Array([
+      Math.random(), Math.random(), Math.random(), Math.random()
+    ]);
+    this.rand_preset = new Float32Array([
+      Math.random(), Math.random(), Math.random(), Math.random()
+    ]);
+
+    // const nonUserKeys = this.qs.concat(this.regs, Object.keys(this.mdVS));
+
+    this.preset.init_eqs();
+
+    // qs need to be initialized to there init values every frame
+    this.mdVSQInit = Utils.pickWasm(this.preset.globals, this.qs);
+    this.mdVSRegs = Utils.pickWasm(this.preset.globals, this.regs);
+
+    // XXX - Don't think we need to keep track of user vars at all, they are just part of WASM state
```

Comment:
That's right. Anything that is not a global is internal to Wasm.

Reply:
woooo!